File: python/compute.py
#!/usr/bin/env python3
"""Join the Noetis network as a compute provider — E2E encrypted tasks + attestation."""

from __future__ import annotations

import argparse
import json
import sys
import time
import urllib.error
import urllib.request
from pathlib import Path

ROOT = Path(__file__).resolve().parent
sys.path.insert(0, str(ROOT / "connection-layer"))

from attestation import build_attestation  # noqa: E402
from chain_state import MIN_STAKE  # noqa: E402
from crypto_wallet import get_or_create_wallet  # noqa: E402
from task_crypto import decrypt_task, generate_enc_keypair  # noqa: E402
from net_utils import get_lan_ip  # noqa: E402
from utils.ollama_client import NOETI_SYSTEM_PREAMBLE, OllamaClient, OllamaError  # noqa: E402

DEFAULT_HUB = "http://127.0.0.1:5052"
POLL_INTERVAL = 1.0
ENC_DIR = ROOT / "connection-layer" / "data" / "wallets"


def api_call(hub: str, method: str, path: str, body: dict | None = None) -> dict:
    url = f"{hub.rstrip('/')}{path}"
    data = json.dumps(body).encode("utf-8") if body is not None else None
    request = urllib.request.Request(
        url,
        data=data,
        headers={"Content-Type": "application/json"},
        method=method,
    )
    try:
        with urllib.request.urlopen(request, timeout=30) as response:
            return json.loads(response.read().decode("utf-8"))
    except urllib.error.HTTPError as exc:
        detail = exc.read().decode("utf-8", errors="replace")
        raise RuntimeError(f"Hub error {exc.code}: {detail}") from exc
    except urllib.error.URLError as exc:
        raise RuntimeError(f"Cannot reach hub at {hub}") from exc


def enc_key_path(node_id: str) -> Path:
    return ENC_DIR / f"enc-{node_id}.json"


def load_enc_keys(node_id: str) -> tuple[str, str]:
    path = enc_key_path(node_id)
    if path.exists():
        payload = json.loads(path.read_text())
        return payload["enc_pubkey"], payload["enc_privkey"]
    pub, priv = generate_enc_keypair()
    ENC_DIR.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps({"enc_pubkey": pub, "enc_privkey": priv}, indent=2))
    path.chmod(0o600)
    return pub, priv


def ensure_stake(hub: str, wallet_name: str, node_id: str) -> tuple[str, object]:
    wallet = get_or_create_wallet(wallet_name)
    status = api_call(
        hub,
        "GET",
        f"/api/staking/status?address={wallet.address}&node_id={node_id}",
    )
    if status.get("eligible"):
        return wallet.address, wallet

    balance = api_call(hub, "GET", f"/api/wallet/balance?address={wallet.address}")
    total = float(balance.get("total", 0))
    if total < MIN_STAKE:
        try:
            api_call(hub, "POST", "/api/faucet", {"address": wallet.address, "amount": 100})
            print(f"[wallet] Faucet credited MLC to {wallet.address}")
        except Exception as exc:
            detail = str(exc)
            if "disabled" in detail.lower() or "faucet" in detail.lower():
                print(
                    f"ERROR: Need {MIN_STAKE} MLC staked — transfer MLC or earn; faucet is off.\n"
                    f"  python3 wallet_cli.py create --name {wallet_name}\n"
                    f"  python3 wallet_cli.py stake --hub {hub} --node-id {node_id}"
                )
            else:
                print(
                    f"ERROR: Need {MIN_STAKE} MLC staked — transfer MLC or earn; faucet is off.\n"
                    f"  ({detail})"
                )
            sys.exit(1)

    nonce = int(api_call(hub, "GET", f"/api/wallet/nonce?address={wallet.address}").get("nonce", 0))
    stake_tx = wallet.sign_transaction(
        {
            "type": "stake",
            "from": wallet.address,
            "amount": MIN_STAKE,
            "node_id": node_id,
            "nonce": nonce,
            "timestamp": time.time(),
        }
    )
    api_call(hub, "POST", "/api/transfer", stake_tx)
    print(f"[wallet] Staked {MIN_STAKE} MLC on-chain")
    return wallet.address, wallet


def main() -> None:
    parser = argparse.ArgumentParser(description="Join Noetis network as compute")
    parser.add_argument("--hub", default=DEFAULT_HUB, help="Network hub URL")
    parser.add_argument("--id", dest="node_id", required=True, help="Your compute node name")
    parser.add_argument("--model", default="qwen2.5:0.5b", help="Ollama model")
    parser.add_argument("--wallet", default="", help="Wallet name (default: compute-<node_id>)")
    parser.add_argument("--access-url", default="", help="Public access URL for this node (shown on entry.txt)")
    args = parser.parse_args()

    client = OllamaClient(model=args.model)
    if not client.is_available():
        print("ERROR: Ollama not running. Install https://ollama.com and run: ollama serve")
        sys.exit(1)

    try:
        model = client.resolve_model([args.model, "qwen2.5:0.5b", "llama3.2:1b"])
    except OllamaError as exc:
        print(f"ERROR: {exc}")
        sys.exit(1)

    wallet_name = args.wallet or f"compute-{args.node_id}"
    wallet_address, wallet = ensure_stake(args.hub, wallet_name, args.node_id)
    enc_pubkey, enc_privkey = load_enc_keys(args.node_id)

    print(f"Compute node: {args.node_id}")
    print(f"Wallet:       {wallet_address}")
    print(f"E2E encrypt:  {enc_pubkey[:20]}…")
    print(f"Hub:          {args.hub}")
    print(f"Model:        {model}")
    print("Waiting for inference tasks… (Ctrl+C to stop)\n", flush=True)

    access_url = args.access_url.strip() or f"http://{get_lan_ip()}:11434"

    api_call(
        args.hub,
        "POST",
        "/api/compute/register",
        {
            "node_id": args.node_id,
            "model": model,
            "wallet_address": wallet_address,
            "enc_pubkey": enc_pubkey,
            "access_url": access_url,
        },
    )

    def fetch_task() -> dict | None:
        # Mesh-first: open offers → claim, then poll fallback.
        try:
            offers_payload = api_call(
                args.hub, "GET", f"/api/compute/offers?node_id={args.node_id}"
            )
            offers = offers_payload.get("offers") or []
            if offers:
                tid = str(offers[0].get("task_id", "")).strip()
                if tid:
                    claimed = api_call(
                        args.hub,
                        "POST",
                        "/api/compute/claim",
                        {"node_id": args.node_id, "task_id": tid},
                    )
                    if claimed and claimed.get("task_id"):
                        return claimed
        except Exception as exc:
            print(f"[warn] offers/claim: {exc}")
        task = api_call(args.hub, "GET", f"/api/compute/poll?node_id={args.node_id}")
        return task if task and task.get("task_id") else None

    while True:
        try:
            api_call(args.hub, "POST", "/api/compute/heartbeat", {"node_id": args.node_id})
            task = fetch_task()
            if task and task.get("task_id"):
                task_id = task["task_id"]
                prompt = decrypt_task(task, enc_privkey)
                prompt_hash = task.get("prompt_hash", "")
                num_predict = task.get("num_predict")
                if num_predict is None:
                    num_predict = task.get("max_tokens")
                try:
                    num_predict = int(num_predict) if num_predict is not None else None
                    if num_predict is not None:
                        num_predict = max(128, min(2048, num_predict))
                except (TypeError, ValueError):
                    num_predict = None
                print(f"[task] {task_id}: {prompt[:80]}…")
                started = time.perf_counter()
                result = client.generate(prompt, system=NOETI_SYSTEM_PREAMBLE, num_predict=num_predict)
                elapsed = (time.perf_counter() - started) * 1000
                attestation = build_attestation(
                    wallet,
                    task_id=task_id,
                    model=result.model,
                    response=result.response,
                    inference_ms=elapsed,
                    prompt_hash=prompt_hash,
                )
                body = {
                    "task_id": task_id,
                    "node_id": args.node_id,
                    "response": result.response,
                    "inference_ms": elapsed,
                    "model": result.model,
                    "attestation": attestation,
                }
                if task.get("encrypted") and task.get("ephem_pubkey"):
                    from task_crypto import encrypt_response

                    body.update(encrypt_response(result.response, task["ephem_pubkey"], enc_privkey))
                    body["response"] = ""
                api_call(args.hub, "POST", "/api/compute/result", body)
                print(f"[done] {task_id} in {elapsed:.0f}ms (attested)")
        except KeyboardInterrupt:
            print("\nCompute node stopped.")
            break
        except Exception as exc:
            print(f"[warn] {exc}")
        time.sleep(POLL_INTERVAL)


if __name__ == "__main__":
    main()

File: rust/crates/noetis-network/src/bin/compute.rs
//! Noetis compute node (Rust) — E2E encrypted tasks + signed attestations.
//! Interoperates with Python hubs; port of `compute.py`.
//!
//! On graceful exit (Ctrl+C / normal return), unregisters from the hub so the
//! mesh slot clears immediately. SIGKILL cannot run Drop — the parent app must
//! call leave_network in that case.

use clap::Parser;
use noetis_chain::attestation::build_attestation;
use noetis_chain::state::MIN_STAKE;
use noetis_chain::taskcrypto::{decrypt_task, encrypt_response, generate_enc_keypair};
use noetis_chain::wallet::Wallet;
use noetis_network::httpc::{get_json, post_json};
use noetis_network::ollama::OllamaClient;
use noetis_network::{data_dir, lan_ip};
use serde_json::{json, Value};
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::time::Duration;

/// POSTs `/api/compute/unregister` when dropped after a successful register.
struct UnregisterGuard {
    hub: String,
    node_id: String,
}

impl Drop for UnregisterGuard {
    fn drop(&mut self) {
        // Best-effort leave — hub may already be gone.
        match post_json(
            &self.hub,
            "/api/compute/unregister",
            &json!({ "node_id": self.node_id }),
            8,
        ) {
            Ok(_) => println!("[unregister] left network as {}", self.node_id),
            Err(e) => eprintln!("[unregister] warn: {e}"),
        }
    }
}

#[derive(Parser)]
#[command(name = "noetis-compute", about = "Join the Noetis network as a compute provider")]
struct Args {
    #[arg(long, default_value = "http://127.0.0.1:5052")]
    hub: String,
    #[arg(long = "id")]
    node_id: String,
    #[arg(long, default_value = "qwen2.5:0.5b")]
    model: String,
    #[arg(long, default_value = "")]
    wallet: String,
    #[arg(long, default_value = "")]
    access_url: String,
}

fn load_enc_keys(node_id: &str) -> (String, String) {
    let dir = data_dir().join("wallets");
    let path = dir.join(format!("enc-{node_id}.json"));
    if let Ok(raw) = std::fs::read_to_string(&path) {
        if let Ok(value) = serde_json::from_str::<Value>(&raw) {
            if let (Some(pub_key), Some(priv_key)) = (
                value.get("enc_pubkey").and_then(Value::as_str),
                value.get("enc_privkey").and_then(Value::as_str),
            ) {
                return (pub_key.to_string(), priv_key.to_string());
            }
        }
    }
    let (pub_key, priv_key) = generate_enc_keypair();
    let _ = std::fs::create_dir_all(&dir);
    let _ = std::fs::write(
        &path,
        serde_json::to_string_pretty(&json!({"enc_pubkey": pub_key, "enc_privkey": priv_key})).unwrap(),
    );
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        let _ = std::fs::set_permissions(&path, std::fs::Permissions::from_mode(0o600));
    }
    (pub_key, priv_key)
}

fn ensure_stake(hub: &str, wallet_name: &str, node_id: &str) -> Result<Wallet, String> {
    let wallet = Wallet::get_or_create(&data_dir().join("wallets"), wallet_name);
    let status = get_json(
        hub,
        &format!(
            "/api/staking/status?address={}&node_id={node_id}",
            wallet.address
        ),
        30,
    )?;
    if status
        .get("eligible")
        .and_then(Value::as_bool)
        .unwrap_or(false)
    {
        println!("[wallet] already eligible · {}", wallet.address);
        return Ok(wallet);
    }

    let balance = get_json(
        hub,
        &format!("/api/wallet/balance?address={}", wallet.address),
        30,
    )?;
    let mut total = balance
        .get("total")
        .and_then(Value::as_f64)
        .unwrap_or(0.0);
    if total < MIN_STAKE {
        match post_json(
            hub,
            "/api/faucet",
            &json!({"address": wallet.address, "amount": 100}),
            30,
        ) {
            Ok(_) => {
                println!("[wallet] Faucet credited MLC to {}", wallet.address);
                std::thread::sleep(std::time::Duration::from_millis(400));
                let balance = get_json(
                    hub,
                    &format!("/api/wallet/balance?address={}", wallet.address),
                    30,
                )?;
                total = balance
                    .get("total")
                    .and_then(Value::as_f64)
                    .unwrap_or(0.0);
            }
            Err(e) => {
                return Err(format!(
                    "Need {MIN_STAKE} MLC to stake for node `{node_id}` — transfer MLC or earn; faucet is off. ({e})"
                ));
            }
        }
    }
    if total < MIN_STAKE {
        return Err(format!(
            "Wallet {} still has {total} MLC (need {MIN_STAKE}) — transfer MLC or earn; faucet is off.",
            wallet.address
        ));
    }

    let nonce = get_json(
        hub,
        &format!("/api/wallet/nonce?address={}", wallet.address),
        30,
    )?
    .get("nonce")
    .and_then(Value::as_i64)
    .unwrap_or(0);
    let stake_tx = wallet.sign_transaction(&json!({
        "type": "stake",
        "from": wallet.address,
        "amount": MIN_STAKE,
        "node_id": node_id,
        "nonce": nonce,
        "timestamp": noetis_chain::now(),
    }));
    post_json(hub, "/api/transfer", &stake_tx, 30)
        .map_err(|e| format!("stake tx failed for node `{node_id}`: {e}"))?;
    println!("[wallet] Staked {MIN_STAKE} MLC on-chain for {node_id}");

    // Confirm eligibility so register won't bounce
    std::thread::sleep(std::time::Duration::from_millis(400));
    let status = get_json(
        hub,
        &format!(
            "/api/staking/status?address={}&node_id={node_id}",
            wallet.address
        ),
        30,
    )?;
    if !status
        .get("eligible")
        .and_then(Value::as_bool)
        .unwrap_or(false)
    {
        return Err(format!(
            "Stake submitted but hub says not eligible yet for `{node_id}`. Wait a few seconds and retry Earn."
        ));
    }
    Ok(wallet)
}

fn main() {
    let args = Args::parse();
    let mut client = OllamaClient::new(&args.model);
    if !client.is_available() {
        eprintln!("ERROR: Ollama not running. Install https://ollama.com and run: ollama serve");
        std::process::exit(1);
    }
    let model = match client.resolve_model(&[&args.model, "qwen2.5:0.5b", "llama3.2:1b"]) {
        Ok(m) => m,
        Err(error) => {
            eprintln!("ERROR: {error}");
            std::process::exit(1);
        }
    };

    let wallet_name = if args.wallet.is_empty() {
        format!("compute-{}", args.node_id)
    } else {
        args.wallet.clone()
    };
    let wallet = match ensure_stake(&args.hub, &wallet_name, &args.node_id) {
        Ok(w) => w,
        Err(error) => {
            eprintln!("ERROR: {error}");
            std::process::exit(1);
        }
    };
    let (enc_pubkey, enc_privkey) = load_enc_keys(&args.node_id);

    println!("Compute node: {}", args.node_id);
    println!("Wallet:       {}", wallet.address);
    println!("E2E encrypt:  {}…", &enc_pubkey[..20.min(enc_pubkey.len())]);
    println!("Hub:          {}", args.hub);
    println!("Model:        {model}");
    println!("Waiting for inference tasks… (Ctrl+C to stop)\n");

    let access_url = if args.access_url.trim().is_empty() {
        format!("http://{}:11434", lan_ip())
    } else {
        args.access_url.trim().to_string()
    };

    if let Err(error) = post_json(
        &args.hub,
        "/api/compute/register",
        &json!({
            "node_id": args.node_id,
            "model": model,
            "wallet_address": wallet.address,
            "enc_pubkey": enc_pubkey,
            "access_url": access_url,
        }),
        30,
    ) {
        eprintln!("ERROR: register failed: {error}");
        std::process::exit(1);
    }
    println!("[register] ok — listed on hub as {}", args.node_id);

    // Unregister on graceful shutdown (Drop). Ctrl+C breaks the loop so Drop runs.
    // Note: SIGKILL cannot run Drop — parent app leave_network must clear the slot.
    let _leave = UnregisterGuard {
        hub: args.hub.clone(),
        node_id: args.node_id.clone(),
    };

    let running = Arc::new(AtomicBool::new(true));
    {
        let flag = running.clone();
        if let Err(e) = ctrlc::set_handler(move || {
            // Exit the poll loop so `_leave` Drop POSTs unregister (process::exit skips Drop).
            flag.store(false, Ordering::SeqCst);
        }) {
            eprintln!("[warn] ctrlc handler not installed: {e}");
        }
    }

    while running.load(Ordering::SeqCst) {
        let cycle = || -> Result<(), String> {
            post_json(&args.hub, "/api/compute/heartbeat", &json!({"node_id": args.node_id}), 30)?;
            // Mesh-first: offers → claim, then poll fallback.
            let mut task = json!({});
            if let Ok(offers_payload) = get_json(
                &args.hub,
                &format!("/api/compute/offers?node_id={}", args.node_id),
                30,
            ) {
                if let Some(offers) = offers_payload.get("offers").and_then(Value::as_array) {
                    if let Some(first) = offers.first() {
                        if let Some(tid) = first.get("task_id").and_then(Value::as_str) {
                            if let Ok(claimed) = post_json(
                                &args.hub,
                                "/api/compute/claim",
                                &json!({"node_id": args.node_id, "task_id": tid}),
                                30,
                            ) {
                                task = claimed;
                            }
                        }
                    }
                }
            }
            if task.get("task_id").and_then(Value::as_str).is_none() {
                task = get_json(
                    &args.hub,
                    &format!("/api/compute/poll?node_id={}", args.node_id),
                    30,
                )?;
            }
            let Some(task_id) = task.get("task_id").and_then(Value::as_str).map(str::to_string) else {
                return Ok(());
            };
            let prompt = decrypt_task(&task, &enc_privkey).ok_or("Failed to decrypt task")?;
            let prompt_hash = task.get("prompt_hash").and_then(Value::as_str).unwrap_or("").to_string();
            let num_predict = task
                .get("num_predict")
                .or_else(|| task.get("max_tokens"))
                .and_then(Value::as_i64)
                .map(|n| n.clamp(128, 2048));
            println!("[task] {task_id}: {}…", prompt.chars().take(80).collect::<String>());

            let result = client.generate_with_options(&prompt, num_predict)?;
            let attestation = build_attestation(
                &wallet,
                &task_id,
                &result.model,
                &result.response,
                result.inference_ms,
                &prompt_hash,
            );
            let mut body = json!({
                "task_id": task_id,
                "node_id": args.node_id,
                "response": result.response,
                "inference_ms": result.inference_ms,
                "model": result.model,
                "attestation": attestation,
            });
            let encrypted = task.get("encrypted").and_then(Value::as_bool).unwrap_or(false);
            if encrypted {
                if let Some(hub_ephem) = task.get("ephem_pubkey").and_then(Value::as_str) {
                    if let Some(enc) = encrypt_response(&result.response, hub_ephem, &enc_privkey) {
                        for (key, value) in enc.as_object().cloned().unwrap_or_default() {
                            body[key] = value;
                        }
                        body["response"] = json!("");
                    }
                }
            }
            post_json(&args.hub, "/api/compute/result", &body, 30)?;
            println!("[done] {task_id} in {:.0}ms (attested)", result.inference_ms);
            Ok(())
        };
        if let Err(error) = cycle() {
            eprintln!("[warn] {error}");
        }
        if !running.load(Ordering::SeqCst) {
            break;
        }
        std::thread::sleep(Duration::from_secs(1));
    }
    println!("[stop] shutting down compute…");
}
